ddtrace/opentelemetry/span_test.go

```diff
@@ -22,6 +22,154 @@ import (
+func waitForTestAgent(done chan struct{}, timeOut time.Duration, t *testing.T) {
+	select {
+	case <-time.After(timeOut):
+		t.FailNow()
+	case <-done:
+		break
+	}
+}
+
+// initializes a test trace provider and agent server
+func getTestTracerProvider(payload *string, done chan struct{},
+	env string, service string, t *testing.T) (*TracerProvider, *http.Server) {
+	s, c := httpmem.ServerAndClient(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/v0.4/traces":
+			buf, err := io.ReadAll(r.Body)
+			if err != nil {
+				t.Fail()
+			}
+			*payload = fmt.Sprintf("%s", buf)
+			done <- struct{}{}
+		}
+		w.WriteHeader(200)
+	}))
+	tp := NewTracerProvider(
+		tracer.WithEnv(env),
+		tracer.WithHTTPClient(c),
+		tracer.WithService(service))
+	return tp, s
+}
+
+func TestSpanSetName(t *testing.T) {
+	var payload string
+	done := make(chan struct{})
+
+	tp, s := getTestTracerProvider(&payload, done, "test_env", "test_srv", t)
+	defer s.Close()
+	otel.SetTracerProvider(tp)
+	tr := otel.Tracer("")
+	defer tp.Shutdown()
+	assert := assert.New(t)
+
+	_, sp := tr.Start(context.Background(), "OldName")
+	// SetName sets the name but not the resource in span - bug?
+	sp.SetName("NewName")
+	sp.End()
+
+	tracer.Flush()
+	waitForTestAgent(done, time.Second, t)
+
+	assert.Contains(payload, "NewName")
+	// following assert fails because "resource" in span is not set by SetName
+	// assert.NotContains(payload, "OldName")
+}
+
+func TestSpanEnd(t *testing.T) {
+
+	var payload string
+	done := make(chan struct{})
+
+	tp, s := getTestTracerProvider(&payload, done, "test_env", "test_srv", t)
+	defer s.Close()
+	otel.SetTracerProvider(tp)
+	tr := otel.Tracer("")
+	defer tp.Shutdown()
+	assert := assert.New(t)
+
+	_, sp := tr.Start(context.Background(), "OldName")
+	sp.SetStatus(codes.Error, "error_description")
+	sp.SetAttributes(attribute.String("span_end_key", "span_end_val"))
+	sp.End()
+	// following operations should not be able to modify the Span
+	sp.SetStatus(codes.Ok, "ok_description")
+	sp.SetAttributes(attribute.String("span_end", "after_end"))
+	sp.SetAttributes(attribute.String("key1", "val1"))
+	sp.SetName("NewName")
+
+	tracer.Flush()
+	waitForTestAgent(done, time.Second, t)
+
+	assert.Contains(payload, "error_description")
+	assert.Contains(payload, "span_end_val")
+	assert.Contains(payload, "OldName")
+
+	assert.NotContains(payload, "after_end")
+	assert.NotContains(payload, "key1")
+	assert.NotContains(payload, "val1")
+	assert.NotContains(payload, "NewName")
+}
+
+func TestSpanSetStatus(t *testing.T) {
+	var payload string
+	done := make(chan struct{})
+
+	tp, s := getTestTracerProvider(&payload, done, "test_env", "test_srv", t)
+	defer s.Close()
+	otel.SetTracerProvider(tp)
+	tr := otel.Tracer("")
+	assert := assert.New(t)
+
+	_, sp := tr.Start(context.Background(), "SpanEndTest")
+	sp.SetStatus(codes.Error, "error_description")
+	// following operation should not do anything
```

Comment:
I would change the status codes from 'Error and Unset' combo to 'OK and Error'. Or add this case separately to show. Because it seems obvous that whatever value is higher than Unset, but it's not obvious that Error can't be propagated over OK